rpc/client.go

```diff
@@ -91,219 +97,224 @@ type Client struct {
+	if err != nil {
+		log.Fatal(err)
+	}
+
 	c := &Client{
-		addr:         addr,
-		Ready:        make(chan struct{}),
 		Closed:       make(chan struct{}),
+		addr:         key,
+		tlsConfig:    tlsConfig,
 		clock:        context.localClock,
 		remoteClocks: context.RemoteClocks,
-		cached:       !context.DisableCache,
-	}
-	if !context.DisableCache {
-		clients[c.Addr().String()] = c
 	}
-	clientMu.Unlock()
 
-	context.Stopper.RunWorker(func() {
-		if err := c.connect(opts, context); err != nil {
-			log.Errorf("client %s failed to connect: %v", c.addr, err)
-		} else {
-			// Signal client is ready by closing Ready channel.
-			log.Infof("client %s connected", c.addr)
-			close(c.Ready)
-
-			// Launch periodic heartbeat. This blocks until the client is to be closed.
-			c.runHeartbeat(context.Stopper)
-		}
-		c.Close()
-	})
-	return c
-}
+	c.healthy.Store(make(chan struct{}))
+	// Must store junk in here so `Load()` doesn't blow up
+	c.conn.Store(internalConn{})
+
+	clients[key] = c
 
-// connect dials the connection in a backoff/retry loop.
-func (c *Client) connect(opts *retry.Options, context *Context) error {
-	// Attempt to dial connection.
 	retryOpts := clientRetryOptions
-	if opts != nil {
-		retryOpts = *opts
-	}
 	retryOpts.Stopper = context.Stopper
 
-	for r := retry.Start(retryOpts); r.Next(); {
-		tlsConfig, err := context.GetClientTLSConfig()
-		if err != nil {
-			// Problem loading the TLS config. Retrying will not help.
-			return err
+	context.Stopper.RunWorker(func() {
+		select {
+		case <-context.Stopper.ShouldStop():
+			c.Close()
+		case <-c.Closed:
 		}
+	})
 
-		conn, err := tlsDialHTTP(c.addr.Network(), c.addr.String(), tlsConfig)
-		if err != nil {
-			// Retry if the error is temporary, otherwise fail fast.
-			if t, ok := err.(net.Error); ok && t.Temporary() {
-				if log.V(1) {
-					log.Warning(err)
-				}
-				continue
-			}
-			return err
-		}
+	context.Stopper.RunWorker(func() {
+		c.runHeartbeat(retryOpts)
 
-		c.mu.Lock()
-		c.Client = rpc.NewClientWithCodec(codec.NewClientCodec(conn))
-		c.lAddr = conn.LocalAddr()
-		c.mu.Unlock()
-		if c.lAddr == nil {
-			return errClosed
+		if client := c.conn.Load().(internalConn).client; client != nil {
+			client.Close()
 		}
+	})
 
-		// Ensure at least one heartbeat succeeds before exiting the
-		// retry loop. If it fails, don't retry: The node is probably
-		// dead.
-		if err = c.heartbeat(); err != nil {
-			return err
-		}
+	return c
+}
 
-		return nil
+// Go wraps net/rpc.Client.Go and provides additional cockroach-specific error
+// handling.
+func (c *Client) Go(serviceMethod string, args interface{}, reply interface{}, done chan *rpc.Call) *rpc.Call {
+	newCall := &rpc.Call{
+		ServiceMethod: serviceMethod,
+		Args:          args,
+		Reply:         reply,
+	}
+	if done == nil {
+		newCall.Done = make(chan *rpc.Call, 1)
+	} else {
+		newCall.Done = done
 	}
 
-	return util.Errorf("system is stopping")
-}
+	go func() {
```

Comment:
This wrapper still feels unnecessary to me; why can't we just update everything that knows to retry on a newRPCError to also retry on ErrUnexpectedEOF?